File: C++/teemo-attacking.cpp
//https://leetcode.com/problems/teemo-attacking/
//Difficulty Level: Medium 
//Tags: Arrays
//Time complexity: O(n)
//Space complexity: O(1)
//similar to overlapping subintervals problems

class Solution {
public:
    int findPoisonedDuration(vector<int>& timeSeries, int duration) {
        
        int n = timeSeries.size();
        if(n==0)
            return 0;
        
        int ans = duration;
        
        for(int i=0; i<n-1; i++)
        {
            ans += duration;
            if(timeSeries[i+1] < timeSeries[i] + duration)  //overlapping condition
            {
                ans -= (timeSeries[i] + duration - timeSeries[i+1]);   //the overlapped time subtracted
            }
        }
        
        return ans;
    }
};

File: Python/teemo-attacking.py
# https://leetcode.com/problems/teemo-attacking/
# Difficulty Level: Medium
# Tags: Array
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def findPoisonedDuration(self, timeSeries: List[int], duration: int) -> int:
        
        size = len(timeSeries)
        
        if size == 0:
            return 0
        
        totalTime = duration
        
        for i in range (size - 1):
            
            totalTime += duration
            
            if timeSeries[i+1] < timeSeries[i] + duration:   
                totalTime -= timeSeries[i] + duration - timeSeries[i+1]
                
        return totalTime
            
